bootloader/piboot.go

```diff
@@ -0,0 +1,421 @@
+		cfgFile, prefix, cfgDir, dstDir)
+
+	if err := os.MkdirAll(dstDir, 0755); err != nil {
+		return err
+	}
+	return p.applyConfig(env, cfgFile, prefix, cfgDir, dstDir)
+}
+
+// Writes os_prefix in RPi config.txt or tryboot.txt
+func (p *piboot) writeRPiCfgWithOsPrefix(prefix, inFile, outFile string) error {
+	buf, err := ioutil.ReadFile(inFile)
+	if err != nil {
+		return err
+	}
+
+	lines := strings.Split(string(buf), "\n")
+
+	replaced := false
+	newOsPrefix := "os_prefix=" + prefix + "/"
+	for i, line := range lines {
+		if strings.HasPrefix(line, "os_prefix=") {
+			if replaced {
+				logger.Noticef("unexpected extra os_prefix line: %q", line)
+				lines[i] = "# " + lines[i]
+				continue
+			}
+			lines[i] = newOsPrefix
+			replaced = true
+		}
+	}
+	if !replaced {
+		lines = append(lines, newOsPrefix)
+		lines = append(lines, "")
+	}
+
+	output := strings.Join(lines, "\n")
+	return osutil.AtomicWriteFile(outFile, []byte(output), 0644, 0)
+}
+
+func (p *piboot) writeCmdline(env *ubootenv.Env, defaultsFile, outFile string) error {
+	buf, err := ioutil.ReadFile(defaultsFile)
+	if err != nil {
+		return err
+	}
+
+	lines := strings.Split(string(buf), "\n")
+	cmdline := lines[0]
+
+	mode := env.Get("snapd_recovery_mode")
+	cmdline += " snapd_recovery_mode=" + mode
+	if mode != "run" {
+		cmdline += " snapd_recovery_system=" + env.Get("snapd_recovery_system")
+	}
+	// Signal when we are trying a new kernel
+	kernelStatus := env.Get("kernel_status")
+	if kernelStatus == "try" {
+		cmdline += " kernel_status=trying"
+	}
+	cmdline += "\n"
+
+	logger.Debugf("writing kernel command line to %s", outFile)
+
+	return osutil.AtomicWriteFile(outFile, []byte(cmdline), 0644, 0)
+}
+
+// Configure pi bootloader with a given os_prefix. cfgDir contains the
+// config files, and dstDir is where we will place the kernel command
+// line.
+func (p *piboot) applyConfig(env *ubootenv.Env,
+	configFile, prefix, cfgDir, dstDir string) error {
+
+	cmdlineFile := filepath.Join(dstDir, "cmdline.txt")
+	refCmdlineFile := filepath.Join(cfgDir, "cmdline.txt")
+	currentConfigFile := filepath.Join(cfgDir, "config.txt")
+
+	if err := p.writeCmdline(env, refCmdlineFile, cmdlineFile); err != nil {
+		return err
+	}
+	if err := p.writeRPiCfgWithOsPrefix(prefix, currentConfigFile,
+		filepath.Join(cfgDir, configFile)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *piboot) GetBootVars(names ...string) (map[string]string, error) {
+	env, err := ubootenv.OpenWithFlags(p.envFile(), ubootenv.OpenBestEffort)
+	if err != nil {
+		return nil, err
+	}
+
+	out := make(map[string]string, len(names))
+	for _, name := range names {
+		out[name] = env.Get(name)
+	}
+
+	return out, nil
+}
+
+func (p *piboot) InstallBootConfig(gadgetDir string, blOpts *Options) error {
+	gadgetFile := filepath.Join(gadgetDir, pibootCfgFilename)
+	st, err := os.Stat(gadgetFile)
+	if err != nil {
+		return err
+	}
+	if st.Size() == 0 {
```

Comment:
what's the use case of the gadget shipping this file non-empty and us just copying it as-is from the gadget? 

Note that we had that special code inside UC20 u-boot bootloader really only for the pi specifically on UC20 because the same code needed to handle both u-boot for not UC20 pi and UC20 pi which really were radically different bootloader layouts, so that's why we had that special marker file in the pi 20 gadget, but we have no such problem AFAICT with the piboot bootloader so I don't see why we don't always just write out an empty new file here instead of copying an existing one from the gadget sometimes

Reply:
Done now